File: identity_agent_ui/rust/src/pqc/keri_serialize.rs
//! keri 1.1.17-compatible icp JSON serialization (SerderKERI makify).

use regex::Regex;
use serde::Serialize;

const KERI_VERSION_MAJOR: u8 = 1;
const KERI_VERSION_MINOR: u8 = 0;
pub(crate) const SAID_DUMMY_LEN: usize = 44;

#[derive(Serialize)]
pub(crate) struct AnchorSeal<'a> {
    pub ia: &'a str,
    pub ka: &'a [String],
}

#[derive(Serialize)]
pub(crate) struct IcpWire<'a> {
    pub v: String,
    pub t: &'a str,
    pub d: &'a str,
    pub i: &'a str,
    pub s: &'a str,
    pub kt: &'a str,
    pub k: &'a [String],
    pub nt: &'a str,
    pub n: &'a [String],
    pub bt: &'a str,
    pub b: &'a [serde_json::Value],
    pub c: &'a [serde_json::Value],
    pub a: &'a [AnchorSeal<'a>],
}

pub(crate) fn versify(size: usize) -> String {
    format!(
        "KERI{:x}{:x}JSON{:06x}_",
        KERI_VERSION_MAJOR, KERI_VERSION_MINOR, size
    )
}

fn patch_version_size(raw: &mut Vec<u8>) -> Result<(), String> {
    let re = Regex::new(r"KERI[0-9a-f][0-9a-f]JSON[0-9a-f]{6}_")
        .map_err(|e| format!("regex: {e}"))?;
    let text = std::str::from_utf8(raw).map_err(|e| format!("utf8: {e}"))?;
    let m = re
        .find(text)
        .ok_or_else(|| "keri version string not found".to_string())?;
    let vs = versify(raw.len());
    if vs.len() != m.len() {
        return Err(format!("version length mismatch: {} vs {}", vs.len(), m.len()));
    }
    let (start, end) = (m.start(), m.end());
    raw[start..end].copy_from_slice(vs.as_bytes());
    Ok(())
}

pub(crate) fn serialize_icp_wire(w: &IcpWire<'_>) -> Result<Vec<u8>, String> {
    let wire = IcpWire {
        v: versify(0),
        t: w.t,
        d: w.d,
        i: w.i,
        s: w.s,
        kt: w.kt,
        k: w.k,
        nt: w.nt,
        n: w.n,
        bt: w.bt,
        b: w.b,
        c: w.c,
        a: w.a,
    };
    let mut raw = serde_json::to_vec(&wire).map_err(|e| format!("marshal: {e}"))?;
    patch_version_size(&mut raw)?;
    Ok(raw)
}

/// Returns (raw bytes, said/prefix digest) after keri 1.1.17 SerderKERI makify.
pub(crate) fn makify_icp_wire<'a>(w: &IcpWire<'a>, dummy: &str) -> Result<(Vec<u8>, String), String> {
    let dummied = IcpWire {
        v: versify(0),
        t: w.t,
        d: dummy,
        i: dummy,
        s: w.s,
        kt: w.kt,
        k: w.k,
        nt: w.nt,
        n: w.n,
        bt: w.bt,
        b: w.b,
        c: w.c,
        a: w.a,
    };
    let raw = serialize_icp_wire(&dummied)?;
    let digest = super::cesr::blake3_qb64(&raw)?;

    let final_wire = IcpWire {
        v: versify(0),
        t: w.t,
        d: &digest,
        i: &digest,
        s: w.s,
        kt: w.kt,
        k: w.k,
        nt: w.nt,
        n: w.n,
        bt: w.bt,
        b: w.b,
        c: w.c,
        a: w.a,
    };
    let raw = serialize_icp_wire(&final_wire)?;
    Ok((raw, digest))
}

File: identity-agent-core/iacrypto/keri_serialize.go
package iacrypto

import (
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
)

const (
	keriVersionMajor = 1
	keriVersionMinor = 0
	keriVersionKind  = "JSON"
	saidDummyLen     = 44 // Blake3_256 fs — Serder.Dummy × fs
)

// anchorSeal is the IA-HYBRID-1 typed seal in standard a anchor.
type anchorSeal struct {
	Ia string   `json:"ia"`
	Ka []string `json:"ka"`
}

// icpWire matches keri 1.1.17 SerderKERI icp field order (v,t,d,i,s,kt,k,nt,n,bt,b,c,a).
type icpWire struct {
	V  string        `json:"v"`
	T  string        `json:"t"`
	D  string        `json:"d"`
	I  string        `json:"i"`
	S  string        `json:"s"`
	Kt string        `json:"kt"`
	K  []string      `json:"k"`
	Nt string        `json:"nt"`
	N  []string      `json:"n"`
	Bt string        `json:"bt"`
	B  []interface{} `json:"b"`
	C  []interface{} `json:"c"`
	A  []anchorSeal  `json:"a"`
	// Di names the delegator, and exists only on a delegated inception.
	//
	// Last, and omitted when empty, because that is where this field sits in
	// the event a delegated inception produces and because an ordinary
	// inception has to keep serializing to exactly the bytes it did before —
	// the identifier is derived from those bytes, so a field appearing in them
	// would change every identifier already created.
	Di string `json:"di,omitempty"`
}

var keriVersionRE = regexp.MustCompile(`KERI[0-9a-f][0-9a-f]JSON[0-9a-f]{6}_`)

func versify(size int) string {
	return fmt.Sprintf("KERI%x%x%s%06x_", keriVersionMajor, keriVersionMinor, keriVersionKind, size)
}

func patchVersionSize(raw []byte) ([]byte, error) {
	loc := keriVersionRE.FindIndex(raw)
	if loc == nil {
		return nil, fmt.Errorf("keri version string not found in raw")
	}
	vs := versify(len(raw))
	if len(vs) != loc[1]-loc[0] {
		return nil, fmt.Errorf("version string length mismatch: got %d want %d", len(vs), loc[1]-loc[0])
	}
	out := make([]byte, len(raw))
	copy(out, raw)
	copy(out[loc[0]:loc[1]], []byte(vs))
	return out, nil
}

func serializeICPWire(w icpWire) ([]byte, error) {
	w.V = versify(0)
	raw, err := json.Marshal(w)
	if err != nil {
		return nil, err
	}
	return patchVersionSize(raw)
}

// makifyICPWire applies keri 1.1.17 SerderKERI makify (dummy d/i, Blake3, same digest).
func makifyICPWire(w icpWire) (icpWire, []byte, error) {
	dummy := strings.Repeat("#", saidDummyLen)
	dummied := w
	dummied.D = dummy
	dummied.I = dummy

	raw, err := serializeICPWire(dummied)
	if err != nil {
		return icpWire{}, nil, err
	}
	dig, err := Blake3QB64(raw)
	if err != nil {
		return icpWire{}, nil, err
	}

	final := w
	final.D = dig
	final.I = dig
	raw, err = serializeICPWire(final)
	if err != nil {
		return icpWire{}, nil, err
	}
	final.V = versify(len(raw))
	return final, raw, nil
}
